### DISCUSSION TITLE
PyVista plot with Solara app in Docker container

### DISCUSSION THREAD

**Discussion by LeDanieI** [pyvista/pyvista#6569] [Q&A] [answered]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
# file: test.py
import pyvista as pv
import ipywidgets as wg
from pyvista import examples

pv.global_theme.trame.jupyter_extension_enabled = False

class Test():
    def __init__(self):
        self.output = wg.Output()
        self.button = wg.Button(description="Show")
        self.button.on_click(self.button_click)
        self.combined = wg.VBox([self.output, self.button])
        # pv.start_xvfb() # Run this only in Docker
        pv.set_jupyter_backend('server')
                
    def button_click(self, event):
        self.show()
        
    def show(self):
        with self.output:
            self.output.clear_output(wait=True) 
            mesh = examples.download_nefertiti()
            mesh.rotate_x(-90.0, inplace=True)
            p = pv.Plotter(notebook=True)
            p.add_mesh(mesh, color="linen", pbr=True, metallic=0.8, roughness=0.1, diffuse=1)
            p.show()
            
    def widget(self):
        return self.combined
        
test = Test()
page = test.widget()
```

### Description

I’m new to both PyVista and Solara and am working on creating a Solara app that incorporates ipywidgets and a PyVista plot. Everything functions correctly when I run it locally on my Windows PC _solara run test.py_. When I click the Show button-widget, the PyVista plot renders in the output-widget.

However, I’m having trouble getting this to work in my Docker container (it does work as a solo widget within Jupyter Lab of the container though). I'm encountering an error (see below). Can anyone advise on how to resolve this issue? I provided a sample code.

### System Information

```shell
ERROR:root:Exception raised by task = <Task finished name='Task-2' coro=<WebAppServer.start() done, defined at /opt/conda/lib/python3.10/site-packages/wslink/backends/aiohttp/__init__.py:117> exception=RuntimeError('set_wakeup_fd only works in main thread of the main interpreter')>
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/asyncio/unix_events.py", line 105, in add_signal_handler
    signal.set_wakeup_fd(self._csock.fileno())
ValueError: set_wakeup_fd only works in main thread of the main interpreter

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/trame_server/core.py", line 710, in on_done
    task.result()
  File "/opt/conda/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception.with_traceback(self._exception_tb)
  File "/opt/conda/lib/python3.10/asyncio/tasks.py", line 232, in __step
    result = coro.send(None)
  File "/opt/conda/lib/python3.10/site-packages/wslink/backends/aiohttp/__init__.py", line 123, in start
    await self._runner.setup()
  File "/opt/conda/lib/python3.10/site-packages/aiohttp/web_runner.py", line 281, in setup
    loop.add_signal_handler(signal.SIGINT, _raise_graceful_exit)
  File "/opt/conda/lib/python3.10/asyncio/unix_events.py", line 107, in add_signal_handler
    raise RuntimeError(str(exc))
RuntimeError: set_wakeup_fd only works in main thread of the main interpreter
```

**Comment by jourdain** [comment]
Wondering if solara is triggering things from a different thread or it is a bug related to Windows even though this only happen when running within docker (linux).

**Reply by banesullivan** [reply]
cc @maartenbreddels, this would be really cool if we could figure out how to combine PyVista/Trame with Solara. Any tips to help us understand how Solara is handling this widget?

**Comment by maartenbreddels** [comment]
Hi 👋 ,

I cannot get the above example to run.
I first get:
```
  File "/Users/maartenbreddels/miniconda3/envs/dev/lib/python3.9/site-packages/pyvista/jupyter/__init__.py", line 48, in _validate_jupyter_backend
    raise ImportError('Please install `trame` and `ipywidgets` to use this feature.')
ImportError: Please install `trame` and `ipywidgets` to use this feature.
```

This is an error in the code, so changing:
```
    if backend in ['server', 'client', 'trame', 'html']:
        try:
            from pyvista.trame.jupyter import show_trame
        except ImportError:  # pragma: no cover
            raise ImportError('Please install `trame` and `ipywidgets` to use this feature.')
```

to
```
    if backend in ['server', 'client', 'trame', 'html']:
        try:
            from pyvista.trame.jupyter import show_trame
        except ModuleNotFoundError:  # pragma: no cover
            raise ImportError('Please install `trame` and `ipywidgets` to use this feature.')
```

I get further, but then I get:
```
  File "/Users/maartenbreddels/miniconda3/envs/dev/lib/python3.9/site-packages/pyvista/trame/__init__.py", line 9, in <module>
    from pyvista.trame.jupyter import elegantly_launch
  File "/Users/maartenbreddels/miniconda3/envs/dev/lib/python3.9/site-packages/pyvista/trame/jupyter.py", line 11, in <module>
    from trame.widgets import vtk as vtk_widgets
ImportError: cannot import name 'vtk' from 'trame.widgets' (/Users/maartenbreddels/miniconda3/envs/dev/lib/python3.9/site-packages/trame/widgets/__init__.py)
```

This is in my messy dev environment, but I usually refrain from ignoring it and starting a new environment from scratch, because there usually is signal in these issues.

Sorry to go off topic first,

Maarten

**Reply by jourdain** [reply]
When pyvista is aimed to be used with jupyter/trame, you need to install it with its optional dependencies [`pyvista[jupyter]`](https://github.com/pyvista/pyvista/blob/main/pyproject.toml#L50-L59). 
If not, you can do it by hand `pip install trame trame-vtk trame-vuetify ipywidgets nest_asyncio`

**Reply by maartenbreddels** [reply]
Great, that downgraded some packages, and now it's showing the original issue (set_wakeup_fd only works in main thread of the main interpreter).

**Comment by maartenbreddels** [comment]
On OSX, this gets me slightly further:

```python
# file: test.py
import pyvista as pv
import ipywidgets as wg
from pyvista import examples

pv.global_theme.trame.jupyter_extension_enabled = False

from pyvista.trame.jupyter import elegantly_launch
from pyvista.trame.views import get_server
import pyvista

server = get_server(name=pyvista.global_theme.trame.jupyter_server_name)

elegantly_launch(server)


class Test():
    def __init__(self):
        self.output = wg.Output()
        self.button = wg.Button(description="Show")
        self.button.on_click(self.button_click)
        self.combined = wg.VBox([self.output, self.button])
        # pv.start_xvfb() # Run this only in Docker
        pv.set_jupyter_backend('server')
                
    def button_click(self, event):
        self.show()
        
    def show(self):
        with self.output:
            self.output.clear_output(wait=True) 
            mesh = examples.download_nefertiti()
            mesh.rotate_x(-90.0, inplace=True)
            p = pv.Plotter(notebook=True)
            p.add_mesh(mesh, color="linen", pbr=True, metallic=0.8, roughness=0.1, diffuse=1)
            p.show()
            
    def widget(self):
        return self.combined
        
test = Test()
page = test.widget()
```
This script makes sure the server runs on the main thread

But then OSX I get:
```
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow should only be instantiated on the main thread!'
```

Curious why this needs the trame server and if we simpler solution could be made that only uses solara/ipywidgets.

**Reply by jourdain** [reply]
The error is related to the fact that the instantiation of a 3D window (OpenGL context) needs to happen on the main thread. There is no work around that since it is an os limitation. 

For skipping the trame server, that is not really possible since no comm is happening over ipywidgets. Trame doesn't have any tight to any Jupyter infrastructure.

**Comment by maartenbreddels** [comment, marked as answer]
Got it working 🥳 
<img width="977" alt="image" src="https://github.com/user-attachments/assets/4dfa9506-096f-4a82-9688-ac104650fde2">

```
SOLARA_KERNEL_THREADED=False solara run app.py --loop asyncio
```

This will not use separate threads for each kernel, which is what we use at https://py.cafe since pyodide does not support threading, however, in this case it's a nice workaround :)
```

**Reply by banesullivan** [reply]
So awesome! I look forward to trying out PyVista + Solara!
